### ISSUE TITLE
Lines payload stream

### ISSUE THREAD

**Issue by fafhrd91** [actix/actix-web#275]
at the moment we have several method how we can read request payload, 
i.e. `HttpMessage::body()`, `HttpMessage::json()` etc. I think we need to support `readlines`

**Comment by ghost** [comment]
how can I do this. Does this mean that HttpMessage::readlines() should return an iterator over strings

**Comment by fafhrd91** [comment]
`::readline()` should return `Stream<Item=String, Error=ReadlinesError>`

**Comment by fafhrd91** [comment]
implementation could be based on `String` extractor.

**Comment by ghost** [comment]
I see that self is `Stream<Item = Bytes, Error = PayloadError>` so should I call map or fold or something like that to accumulate bytes until newline character than make a string and return it?

**Comment by fafhrd91** [comment]
yes. just add config about line length. also you should handle encoding properly.

**Comment by ghost** [comment]
ok. I am tired and this seems a little hard I'll try it tomorrow

**Comment by ghost** [comment]
```rust
struct ReadLines<T>
where
    Self: Stream<Item = Bytes, Error = PayloadError> + Sized,
{
    from: T,
    buff: String,
}

impl<T> ReadLines<T>
where
    Self: Stream<Item = Bytes, Error = PayloadError> + Sized,
{
    fn new(from: T) -> Self {
        ReadLines {
            from,
            buff: String::new(),
        }
    }
}

impl<T> Stream for ReadLines<T>
where
    Self: Stream<Item = Bytes, Error = PayloadError> + Sized,
{
    type Item = String;
    type Error = ReadlinesError;
    
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        // do stuff
    }
}
```
is this the right way?

**Comment by fafhrd91** [comment]
Mostly. 

1. I don’t think you need T
2. take stream ownership
3. You can not really use String as buffer because you can get bytes chunk splinted in the middle of Unicode codepoint. Use BytesMut instead.

**Comment by ghost** [comment]
I got this now
```
struct ReadLines {
    from: ???,
    buff: BytesMut,
}

impl ReadLines {
    fn new(from: ???) -> Self {
        ReadLines {
            from,
            buff: BytesMut::new(),
        }
    }
}

impl Stream for ReadLines {
    type Item = String;
    type Error = ReadlinesError;
    
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        // do stuff
    }
}
```
how can I take ownership of the stream?

**Comment by fafhrd91** [comment]
You can make readline generic over stream. Check httpmesaage.rs

**Comment by ghost** [comment]
Is this somewhat acceptable?
```Rust
struct Readlines<T>
where
    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
{
    req: T,
    buff: Vec<u8>,
}

impl<T> Readlines<T>
where
    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
{
    fn new(req: T) -> Self {
        Readlines {
            req,
            buff: Vec::with_capacity(256),
        }
    }
}

impl<T> Stream for Readlines<T>
where
    T: HttpMessage + Stream<Item = Bytes, Error = PayloadError> + 'static,
{
    type Item = String;
    type Error = ReadlinesError;
    
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        match self.req.poll() {
            Ok(Async::Ready(Some(bytes))) => {
                for b in bytes.iter() {
                    if *b == '\n' as u8 {
                        self.buff.push(*b);
                        let line = str::from_utf8(&*self.buff)?.to_owned();
                        self.buff.clear();
                        return Ok(Async::Ready(Some(line)));
                    } else {
                        self.buff.push(*b);
                    }
                }
                Ok(Async::NotReady)
            },
            Ok(Async::NotReady) => Ok(Async::NotReady),
            Ok(Async::Ready(None)) => {
                if self.buff.len() == 0 {
                    return Ok(Async::Ready(None));
                }
                let line = str::from_utf8(&*self.buff)?.to_owned();
                self.buff.clear();
                return Ok(Async::Ready(Some(line)))
            },
            Err(e) => Err(ReadlinesError::from(e)),
        }
    }
}

pub enum ReadlinesError {
    EncodingError,
    PayloadError(PayloadError),
}

impl From<PayloadError> for ReadlinesError {
    fn from(err: PayloadError) -> Self {
        ReadlinesError::PayloadError(err)
    }
}

impl From<str::Utf8Error> for ReadlinesError {
    fn from(_: str::Utf8Error) -> Self {
        ReadlinesError::EncodingError
    }
}
```
I will add line length limit

**Comment by ghost** [comment]
closed by #312
